### ISSUE TITLE
The Macro Editor does not display Japanese characters.

### ISSUE THREAD

**Issue by marimelon** [daemitus/SomethingNeedDoing#4]
The Macro Editor does not display Japanese characters.

![215157](https://user-images.githubusercontent.com/29140293/115873912-77f03e80-a47e-11eb-9509-a5290a1f206b.jpg)

**Comment by rlynch-ironnet** [comment]
do you know of a monospaced font that does jp symbols?

**Comment by marimelon** [comment]
"Noto Sans Mono CJK JP" is famous.
This width ratio is alphabet : Japanese = 1:2

**Comment by marimelon** [comment]
my idea

fonts : https://noto-website-2.storage.googleapis.com/pkgs/NotoSansCJKjp-hinted.zip

```C#
public void UiBuilder_OnBuildFonts()
{
    var fontData = SomethingNeedDoingPlugin.ReadResource("SomethingNeedDoing.Fonts.ProggyVectorRegular.ttf");
    var fontPtr = Marshal.AllocHGlobal(fontData.Length);
    Marshal.Copy(fontData, 0, fontPtr, fontData.Length);

    var JPfontData = SomethingNeedDoingPlugin.ReadResource("SomethingNeedDoing.Fonts.NotoSansMonoCJKtc-Regular.otf");
    var JPfontPtr = Marshal.AllocHGlobal(JPfontData.Length);
    Marshal.Copy(JPfontData, 0, JPfontPtr, JPfontData.Length);

    unsafe{
        ImFontConfigPtr fontConfig = ImGuiNative.ImFontConfig_ImFontConfig();
        fontConfig.MergeMode = true;
        fontConfig.PixelSnapH = true;

        ProggyVectorFont = ImGui.GetIO().Fonts.AddFontFromMemoryTTF(fontPtr, fontData.Length, plugin.Configuration.CustomFontSize, fontConfig);

        // Set Japanese Font GlyphRanges
        // Copy from
        // https://github.com/goatcorp/Dalamud/blob/master/Dalamud/Interface/GlyphRangesJapanese.cs
        var japaneseRangeHandle = GCHandle.Alloc(GlyphRangesJapanese.GlyphRanges, GCHandleType.Pinned);
        NotoSansMonoFont = ImGui.GetIO().Fonts.AddFontFromMemoryTTF(JPfontPtr, JPfontData.Length, plugin.Configuration.CustomFontSize, fontConfig, japaneseRangeHandle.AddrOfPinnedObject());
        japaneseRangeHandle.Free();
    }
}
```

```C#
// PluginUI.DisplayMacroEdit()
ImGui.PushFont(ProggyVectorFont);
ImGui.PushFont(NotoSansMonoFont);

var contents = node.Contents;
if (ImGui.InputTextMultiline($"##{node.Name}-editor", ref contents, 100_000, new Vector2(-1, -1)))
{
    node.Contents = contents;
    plugin.SaveConfiguration();
}

ImGui.PopFont();
ImGui.PopFont();
```

**Comment by daemitus** [comment]
I assume you meant NotoSansMonoCJKjp-Regular.otf and not tc?

**Comment by marimelon** [comment]
Oh, my mistake.

**Comment by marimelon** [comment]
I'm not familiar with c#, should we release fontPtr (and JPfontPtr) in Dispose?

**Comment by daemitus** [comment]
let me see what dalamud does with the default font ptr

**Comment by daemitus** [comment]
There is no dispose method, I think ImGui handles it.

**Comment by daemitus** [comment]
Implemented in 1.0.0.10.
